HPCC-14417 Implement JSON Fetch

afishbeck opened the pull request:
Also fix JSON BuildIndex virtual filepositions, and Roxie read JSON activity.

Signed-off-by: Anthony Fishbeck anthony.fishbeck@lexisnexis.com

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday Please review

richardkchapman commented on dali/ft/daftformat.ipp:
```diff
@@ -473,27 +474,32 @@ public:
             {
                 footerLength = isRootless() ? 0 : 1; //account for parser using ] as offset unless rootless
-                footerLength = footerLength + size - rowEnd;
+                footerLength = footerLength + size - rowEnd - 1;
```
Why the -1 ? Was this previously a bug?

afishbeck replied:
No, I made the parser more consistent with xml so that the virtual filepositions in my index were right..  that had a side effect of changine the way the offsets had to be handled in spraying, but I think its more correct.

richardkchapman commented on dali/ft/daftformat.ipp:
```diff
@@ -502,12 +508,12 @@ public:
         if (rowStart <= headerLength)
             return 0;
-        return rowStart - headerLength - 1;
+        return rowStart - headerLength;
```
Why?

afishbeck replied:
Same change to the offset to be compatible with how index virtual file positions are set.  Better to make a small change in spraying then to complicate indexing.

richardkchapman commented on ecl/hqlcpp/hqlsource.cpp:
```diff
@@ -7275,6 +7276,8 @@ ABoundActivity * HqlCppTranslator::doBuildActivityFetch(BuildCtx & ctx, IHqlExpr
     case no_xml:
         return info.buildActivity(ctx, expr, TAKxmlfetch, "XmlFetch", childActivity);
+    case no_json:
+        return info.buildActivity(ctx, expr, TAKjsonfetch, "XmlFetch", childActivity);
```
Should this say JsonFetch rather than XmlFetch?

richardkchapman replied:
(if not a comment indicating it is deliberate might be a good idea)

afishbeck replied:
I'll add a comment, because I still want to have the generated code take advantage of the existing xmlFetch classes,

afishbeck commented:
@richardkchapman Updated with comment based on review.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
I didn't spot anything else. @ghalliday Please sanity-check

ghalliday commented:
@afishbeck the code changes look good, but it needs a test added to the regression suite.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

afishbeck committed:
HPCC-14417 Implement JSON Fetch

Also fix JSON BuildIndex virtual filepositions.
Also fix Roxie XML build index filepositions.
Also fix hthor XML fetch (not clearing lastMatch).

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@richardkchapman 
@ghalliday 

Please re-review.  I've refactored, fixed a roxie issue indexing xml, fixed an hthor issue in xml fetch, and added a test case.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Fail | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 
```xml
setupsearchindex on thor failed with:
<Exception><Code>1303</Code><Source>eclagent</Source><Message>System error: 1303: Graph[65], join[73]: Invalid pointer, Master exception</Message></Exception>

```
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| jsonfetch | Pass | Pass | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK

afishbeck committed:
HPCC-14417 Add additional xml fetch test case

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

richardkchapman commented:
@jakesmith @afishbeck Anyone have an explanation for the smoketest failure?

afishbeck commented:
Let me try and reproduce.

afishbeck commented:
@richardkchapman @jakesmith setupsearchindex.ecl passes on ubuntu for my build of this branch.  Any ideas on further testing?

jakesmith commented:
@afishbeck - have you tried running through Valgrind?

richardkchapman commented:
@AttilaVamos Can we tell if the smoketest failure is reproducible?

AttilaVamos commented:
Something strange thing happened, because this error occurred in the third run backward and not in the last. I should check the log processing.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| jsonfetch | Pass | Pass | Pass |
| xmlfetch2 | Pass | Pass | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
I checked the logs of this rescheduled smoketest. This result is reliable and matches the result of setup and two ECL test cases.
Previously there was a problem with GitHub comment insertion. An error message in previous run contained some characters which caused problem of GitHub to process the text of comment.

afishbeck commented:
@richardkchapman I vaguely remember having missed a change when I initially made the "final" commit.  I added the file, but then squashed the last commit, so unfortunately I think the commit that may have failed no longer exists.. and no idea which bit I had left off.

richardkchapman commented:
@ghalliday please re-review

ghalliday commented on ecl/hthor/hthorkey.cpp:
```diff
@@ -2597,7 +2598,7 @@ class XmlFetchPartHandler : public SimpleFetchPartHandlerBase, public IXMLSelect
                 throw owner.makeWrappedException(e, fname.str());
             }
-            if(!gotNext)
+            if(!gotNext && !lastMatch) //unfortunately json parser next() has slightly different behavior, tricky, may fix later
```
This comment doesn't explain how it is different or why the condition needs to check for !lastMatch.  You may not remember if you ever revisit it.
Is this something that is likely to be worth changing?

afishbeck replied:
It's certainly better to emulate the same behavior.  I'll look at what it would take and what the impact might be to other uses of the interface.

ghalliday commented on roxie/ccd/ccdserver.cpp:
```diff
@@ -20594,6 +20594,20 @@ class CRoxieServerXmlReadActivity : public CRoxieServerDiskReadBaseActivity, imp
+    //interface IThorDiskCallback
+    virtual unsigned __int64 getFilePosition(const void * row)
+    {
+        return localOffset;
```
This looks suspicious since it returns the same the the function below.

afishbeck replied:
Agreed.  I'll look at calculating the proper position.

ghalliday commented:
@afishbeck a couple of minor questions/comments.

richardkchapman commented:
@afishbeck Please comment

afishbeck commented on roxie/ccd/ccdactivities.cpp:
```diff
@@ -1608,19 +1608,38 @@ class CsvRecordProcessor : public RecordProcessor
+    }
+    virtual unsigned __int64 getLocalFilePosition(const void * row)
+    {
+        if (!reader || !owner.varFileInfo || !owner.varFileInfo->getFileMap())
```
@richardkchapman can you comment on what scenarios may not have varFileInfo, and any alternative way of getting the part offset?

richardkchapman replied:
varfileinfo means that the file was resolved dynamically when the query was run rather than statically when the query is loaded.

In the latter case you get the file info from the factory.

See eg this snippet:

      rawFile.setown(variableFileName ? varFiles->getFilePart(lastPartNo.partNo, base) : factory->getFilePart(lastPartNo.partNo, base)); // MORE - superfiles

afishbeck replied:
Thanks @richardkchapman , teaches me for running my test in the least typical way.

richardkchapman commented:
@afishbeck back to you

richardkchapman commented:
@afishbeck Back to you again

afishbeck committed:
HPCC-14417 Further JSON parsing and fileposition fixes

Make JSON pull parser interface getNext() compatible with XML pull parser.

Fix roxie filepostion support on XML record handler.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
Updated.

JsonPullParser::getNext() now behaves like xml in terms of EOF handling.

Fixed how roxie fileposition and localfileposition are calculated.
